source/extensions/filters/http/oauth/oauth_client.cc

```diff
@@ -0,0 +1,148 @@
+#include "common/common/logger.h"
+#include "common/http/message_impl.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Oauth {
+
+static const std::string AuthMeApiEndpoint = "/api/me/";
+static const std::string AuthTokenEndpoint = "/oauth/token/";
+
+// JSON keys for the various responses
+static const std::string kAccessToken = "access_token";
+static const std::string kExpiresIn = "expires_in";
+static const std::string kUser = "user";
+static const std::string kUsername = "username";
+
+static const std::string GetAccessTokenBodyFormatString =
+    "grant_type=authorization_code&code={0}&client_id={1}&client_secret={2}&redirect_uri={3}";
+
+void OAuth2ClientImpl::asyncGetAccessToken(const std::string& auth_code,
+                                           const std::string& client_id, const std::string& secret,
+                                           const std::string& cb_url) {
+  Http::RequestMessagePtr request = createPostRequest();
+  request->headers().setPath(AuthTokenEndpoint);
+  const std::string body =
+      fmt::format(GetAccessTokenBodyFormatString, auth_code, client_id, secret, cb_url);
+  request->body() = std::make_unique<Buffer::OwnedImpl>(body);
+
+  ENVOY_LOG(debug, "Dispatching OAuth request for access token.");
+  dispatchRequest(std::move(request));
+  state_ = OAuthState::PendingAccessToken;
+}
+
+void OAuth2ClientImpl::asyncGetIdentity(const std::string& access_token) {
+  Http::RequestMessagePtr request = createAuthGetRequest(access_token);
+  request->headers().setPath(AuthMeApiEndpoint);
+
+  ENVOY_LOG(debug, "Dispatching OAuth request for user identity.");
+  dispatchRequest(std::move(request));
+  state_ = OAuthState::PendingIdentity;
+}
+
+void OAuth2ClientImpl::dispatchRequest(Http::RequestMessagePtr&& msg) {
+  in_flight_request_ = cm_.httpAsyncClientForCluster(cluster_name_)
+                           .send(std::move(msg), *this,
+                                 Http::AsyncClient::RequestOptions().setTimeout(timeout_duration_));
+}
+
+// successful request calls will end up here
+void OAuth2ClientImpl::onSuccess(const Http::AsyncClient::Request&,
+                                 Http::ResponseMessagePtr&& message) {
+  in_flight_request_ = nullptr;
+
+  /**
+   * Due to the asynchonous nature of this client, it's important to immediately update the state
+   * that we are not waiting on anything, otherwise a painful debugging session may ensue.
+   */
+  OAuthState prior_state = state_;
+  state_ = OAuthState::Idle;
+
+  // Check that the auth cluster returned a happy response.
+  const auto response_code = message->headers().Status()->value().getStringView();
+  if (response_code != "200") {
+    ENVOY_LOG(debug, "Oauth response code: {}", response_code);
+    parent_->sendUnauthorizedResponse();
+    return;
+  }
+
+  const std::string response_body = message->bodyAsString();
+
+  Json::ObjectSharedPtr json_object;
```

Comment:
Should we define a protobuf message and just use that JSON marshalling here instead? We can then do the constraints in the `.pb` file and clean up the static string keys here.

Reply:
+1 if we can express this as a proto that would be great